include/swift/AST/DiagnosticsParse.def

```diff
@@ -131,17 +131,17 @@ ERROR(lex_unicode_escape_braces,none,
 ERROR(lex_illegal_multiline_string_start,none,
       "multi-line string literal content must begin on a new line", ())
 ERROR(lex_illegal_multiline_string_end,none,
-      "multi-line string literal content must not be on the delimiter's line", ())
-ERROR(lex_insufficient_string_indent,none,
-      "string literal content is insufficiently indented", ())
-ERROR(lex_unexpected_tab_in_string_indent,none,
-      "string literal content is indented with a tab where a space is expected", ())
-ERROR(lex_unexpected_space_in_string_indent,none,
-      "string literal content is indented with a space where a tab is expected", ())
+      "multi-line string literal closing delimiter must begin on a new line", ())
+ERROR(lex_inconsistent_string_indent,none,
+      "line does not start with same indentation as closing delimiter of multi-line string literal", ())
+NOTE(note_insufficient_string_indent,none,
+      "unexpectedly found non-indentation character here", ())
+NOTE(note_unexpected_char_in_string_indent,none,
+      "unexpected %select{space|tab}0 in indentation here", (bool))
+NOTE(note_matching_indentation_char_here,none,
+      "should match %select{space|tab}0 in same position before closing delimiter", (bool))
```

Comment:
Let's see. I think there are a few separate interesting cases here:

- Every line is consistent except the last one.
- Most lines are consistent with the last one, but a few are not.
- Everything is terrible and we can't tell what the indentation ought to be.

In the first case, the delimiter's clearly the one that's wrong, and it would be nice if we got one error and a note to fix it.

In the second case, we'd like to fix the offending line(s). I agree with Brent that it makes sense to show the final line as well, at least once, and I stand by my earlier comment that simply replacing all leading indentation may not be the correct fix and therefore should be on a note. The thing I might add to that is that if *consecutive* lines have the *same wrong indentation,* it might be nice™ to fix them all at once. But that's kind of a bonus.

The third case is just bad. Here, we could take a hint from Clang and just stop emitting errors after maybe the sixth mismatch. "note: suppressing any remaining indentation errors for this string literal"

None of this is a comment on the *text* of the diagnostic. I think Xiaodi's generally right that the notes can be a bit terser given the context of the error. The error should stand on its own but doesn't have to include the recommended fix.